third_party/tzcode/strftime.c

```diff
@@ -0,0 +1,679 @@
+	const char *mon[MONSPERYEAR];
+	const char *month[MONSPERYEAR];
+	const char *wday[DAYSPERWEEK];
+	const char *weekday[DAYSPERWEEK];
+	const char *X_fmt;
+	const char *x_fmt;
+	const char *c_fmt;
+	const char *am;
+	const char *pm;
+	const char *date_fmt;
+};
+
+#define Locale (&C_time_locale)
+
+static const struct lc_time_T C_time_locale = {
+	{"Jan", "Feb", "Mar", "Apr", "May", "Jun",
+	 "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"},
+	{"January", "February", "March", "April", "May", "June",
+	 "July", "August", "September", "October", "November", "December"},
+	{"Sun", "Mon", "Tue", "Wed",
+	 "Thu", "Fri", "Sat"},
+	{"Sunday", "Monday", "Tuesday", "Wednesday",
+	 "Thursday", "Friday", "Saturday"},
+
+	/* X_fmt */
+	"%H:%M:%S",
+
+	/*
+	 ** x_fmt
+	 ** C99 and later require this format.
+	 ** Using just numbers (as here) makes Quakers happier;
+	 ** it's also compatible with SVR4.
+	 */
+	"%m/%d/%y",
+
+	/*
+	 ** c_fmt
+	 ** C99 and later require this format.
+	 ** Previously this code used "%D %X", but we now conform to C99.
+	 ** Note that
+	 ** "%a %b %d %H:%M:%S %Y"
+	 ** is used by Solaris 2.3.
+	 */
+	"%a %b %e %T %Y",
+
+	/* am */
+	"AM",
+
+	/* pm */
+	"PM",
+
+	/* date_fmt */
+	"%a %b %e %H:%M:%S %Z %Y"
+};
+
+static int pow10[] = { 1,      10,	100,	  1000,	     10000,
+		       100000, 1000000, 10000000, 100000000, 1000000000 };
+
+enum warn {
+	IN_NONE,
+	IN_SOME,
+	IN_THIS,
+	IN_ALL
+};
+
+static char *
+_add(const char *, char *, const char *);
+static char *
+_conv(int, const char *, char *, const char *);
+static char *
+_fmt(const char *, const struct datetime_tm *, char *, const char *, enum warn *);
+static char *
+_yconv(int, int, bool, bool, char *, char const *);
+
+/*
+ * Given the seconds from Epoch (1970-01-01) we calculate date
+ * since Rata Die (0001-01-01).
+ * DT_EPOCH_1970_OFFSET is the distance in days from Rata Die to Epoch.
+ */
+static int
+local_dt(int64_t secs)
+{
+	return dt_from_rdn((int)(secs / SECS_PER_DAY) + DT_EPOCH_1970_OFFSET);
+}
+
+static int64_t
+local_secs(const struct datetime *date)
+{
+	return (int64_t)date->epoch + date->tzoffset * 60;
+}
+
+void
+datetime_to_tm(const struct datetime *date, struct datetime_tm *tm)
+{
+	struct tm t;
+	memset(&t, 0, sizeof(t));
+	tm->tm_epoch = local_secs(date);
+	dt_to_struct_tm(local_dt(tm->tm_epoch), &t);
+	tm->tm_year = t.tm_year;
+	tm->tm_mon = t.tm_mon;
+	tm->tm_mday = t.tm_mday;
+	tm->tm_wday = t.tm_wday;
+	tm->tm_yday = t.tm_yday;
+
+	tm->tm_gmtoff = date->tzoffset * 60;
+	tm->tm_nsec = date->nsec;
+
+	int seconds_of_day = tm->tm_epoch % SECS_PER_DAY;
+	tm->tm_hour = (seconds_of_day / 3600) % 24;
+	tm->tm_min = (seconds_of_day / 60) % 60;
+	tm->tm_sec = seconds_of_day % 60;
+}
+
+size_t
+tnt_strftime(char *s, size_t maxsize, const char *format,
+	     const struct datetime_tm *t)
+{
+	char *p;
+	int saved_errno = errno;
+	enum warn warn = IN_NONE;
+
+	tzset();
+	p = _fmt(format, t, s, s + maxsize, &warn);
+	if (!p) {
+		errno = EOVERFLOW;
```

Comment:
Where does it return `NULL` ever? I don't see. I added `assert(false)` to there and the tests passed. Please, cover this with a test.

Reply:
Refactored (for `SNPRINT` usage) code does not have this line anymore...